src/core/server/application/application_context_provider.ts

```diff
@@ -0,0 +1,175 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { compileFile } from 'pug';
+import { join } from 'path';
+
+import { i18n } from '@kbn/i18n';
+
+import { LegacyServiceSetup } from '../legacy';
+import { Env } from '../config';
+import { InternalHttpServiceSetup, KibanaRequest, KibanaResponseFactory } from '../http';
+import { PluginsServiceSetup } from '../plugins';
+import { createCspRuleString } from '../csp';
+import { mergeCapabilities } from './capabilities';
+import { mergeVariables } from './merge_variables';
+import { IApplicationContextProvider } from './types';
+
+interface Params {
+  request: KibanaRequest;
+  response: KibanaResponseFactory;
+  env: Env;
+  legacy: LegacyServiceSetup;
+  http: InternalHttpServiceSetup;
+  uiSettings: Record<string, any>;
+  plugins: PluginsServiceSetup;
+}
+
+interface PluginSpec {
+  readConfigValue: (config: any, key: any) => any;
+}
+
+interface Provider<T = object> {
+  fn: (
+    server: InternalHttpServiceSetup['server'],
+    config: ReturnType<PluginSpec['readConfigValue']>
+  ) => T;
+  pluginSpec: PluginSpec;
+}
+
+type App = ReturnType<InternalHttpServiceSetup['server']['getUiAppById']>;
+type Replacer<T = object> = (
+  variables: T,
+  request: KibanaRequest,
+  server: InternalHttpServiceSetup['server']
+) => Promise<T> | T;
+
+export class ApplicationContextProvider implements IApplicationContextProvider {
+  private readonly request!: KibanaRequest;
+  private readonly response!: KibanaResponseFactory;
+  private readonly env!: Env;
+  private readonly legacy!: LegacyServiceSetup;
+  private readonly http!: InternalHttpServiceSetup;
+  private readonly uiSettings!: Record<string, any>;
+  private readonly plugins!: PluginsServiceSetup;
+  private readonly template = compileFile(join(__dirname, 'views/ui_app.pug'));
```

Comment:
question: Is it okay that we compile the file on every request? Probably the result is cached. But that's an implementation detail that we cannot rely on.